File: src/main/java/org/elmlang/intellijplugin/features/structure/ElmTypeAliasTreeElement.java
package org.elmlang.intellijplugin.features.structure;

import com.intellij.ide.structureView.StructureViewTreeElement;
import com.intellij.ide.structureView.impl.common.PsiTreeElementBase;
import org.elmlang.intellijplugin.ElmIcons;
import org.elmlang.intellijplugin.psi.ElmTypeAliasDeclaration;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import javax.swing.*;
import java.util.Collection;
import java.util.Collections;
import java.util.Optional;

class ElmTypeAliasTreeElement extends PsiTreeElementBase<ElmTypeAliasDeclaration> {

    ElmTypeAliasTreeElement(ElmTypeAliasDeclaration psiElement) {
        super(psiElement);
    }

    @Override
    public Icon getIcon(boolean open) {
        return ElmIcons.TYPE_ALIAS;
    }

    @Nullable
    @Override
    public String getPresentableText() {
        return Optional.ofNullable(getElement())
                .map(e -> e.getUpperCaseId().getName())
                .orElse(null);
    }

    @NotNull
    @Override
    public Collection<StructureViewTreeElement> getChildrenBase() {
        return Collections.emptyList();
    }
}

File: src/main/java/org/elmlang/intellijplugin/features/structure/ElmTypeAliasTreeElement.kt
package org.elmlang.intellijplugin.features.structure

import com.intellij.ide.structureView.StructureViewTreeElement
import com.intellij.ide.structureView.impl.common.PsiTreeElementBase
import com.intellij.psi.presentation.java.SymbolPresentationUtil
import org.elmlang.intellijplugin.ElmIcons
import org.elmlang.intellijplugin.psi.ElmTypeAliasDeclaration

class ElmTypeAliasTreeElement(element: ElmTypeAliasDeclaration): PsiTreeElementBase<ElmTypeAliasDeclaration>(element) {

    override fun getIcon(open: Boolean) =
            ElmIcons.TYPE_ALIAS

    override fun getPresentableText(): String? {
        val decl = element
        return when (decl) {
            is ElmTypeAliasDeclaration ->
                decl.upperCaseId.name
            else ->
                decl?.let { SymbolPresentationUtil.getSymbolPresentableText(it) }
        }
    }

    override fun getChildrenBase(): Collection<StructureViewTreeElement> =
            emptyList()
}
